### PR TITLE
[WIP] MGMT-4872 - Initial events generation

### PR DESCRIPTION
**Pull request by ronniel1**
The events will be declared in a YAML file. Events Golang code is generated from that YAML.
This allows us to have an ID for events and also to make it easier to review the working of events.
One of the benefits of having a unique ID is that tests (subsystem. UT and E2E) will be able to rely on the event ID and not on the text of the event

### CODE REVIEW

File: tools/generate_events.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+import logging
+import argparse
+import jinja2
+from collections import OrderedDict
+
+
+EVENTS_DEFINITION = 'x-events'
+
+# NOQA
+EVENT_TEMPLATE = jinja2.Template('''// Code generated by from events definition in swagger; DO NOT EDIT.
```
Comment (avishayt) [inline tools/generate_events.py:14]:
```suggestion
EVENT_TEMPLATE = jinja2.Template('''// Code generated by from events definition; DO NOT EDIT.
```

File: tools/generate_events.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    invalid_props = INVALID_HOST_PROPERTIES
+    if e['event_type'] == "cluster":
+        invalid_props = INVALID_CLUSTER_PROPERTIES
+    for p in invalid_props:
+        if p in e['properties']:
+            raise Exception("Invalid '{}' in properties of {}".format(p, e['name']))
+
+
+def parse(swagger_path):
```
Comment (avishayt) [inline tools/generate_events.py:229]:
```suggestion
def parse(yaml_path):
```

File: tools/generate_events.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    def dict_ctor(loader, node):
+        return OrderedDict(loader.construct_pairs(node))
+
+    # Need to use ordered dict with yaml.Loader so that the order of definitions will not change
+    # each time we generate
+    # This is specifically important for the order of the event ctor arguments
+    yaml.Loader.add_constructor(yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG, dict_ctor)
+    with open(swagger_path, "r") as spec_file:
```
Comment (avishayt) [inline tools/generate_events.py:237]:
```suggestion
    with open(yaml_path, "r") as spec_file:
```

File: tools/generate_events.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        spec = yaml.load(spec_file, Loader=yaml.Loader)
+
+    if spec is None:
+        raise ValueError("Invalid input file")
+
+    events_spec = spec.get(EVENTS_DEFINITION)
+    if events_spec is None:
+        logging.error("Cannot find events in %s", swagger_path)
```
Comment (avishayt) [inline tools/generate_events.py:245]:
```suggestion
        logging.error("Cannot find events in %s", yaml_path)
```

File: tools/generate_events.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+    return events, extra_imports
+
+
+def main():
+    """Tool entry point - read parameters from users and start tool."""
+    parser = argparse.ArgumentParser(
+        description='Generate events definitions based on specifications.'
+    )
+    parser.add_argument('source', type=str, help='swagger source path')
```
Comment (avishayt) [inline tools/generate_events.py:269]:
```suggestion
    parser.add_argument('source', type=str, help='yaml source path')
```

File: tools/generate_events.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    parser.add_argument('dest', type=str, help='destination file path')
+    parser.add_argument('-p', '--package-name', default="events", help='go package name')
+    args = parser.parse_args()
+
+    events, extra_imports = parse(swagger_path=args.source)
```
Comment (avishayt) [inline tools/generate_events.py:274]:
```suggestion
    events, extra_imports = parse(yaml_path=args.source)
```

File: tools/generate_events.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        if self.event.severity.lower() == "warning":
+            return "warning"
+        if self.event.severity.lower() == "error":
+            return "error"
+        if self.event.severity.lower() == "critical":
+            return "critical"
+
+    def get_id(self):
+        EventGenerator.id_counter += 1
+        return "%s%04d" % (self.event.id_prefix, self.id_counter)
```
Comment (avishayt) [inline tools/generate_events.py:143]: This means we need to be careful to always append to the file right?

Comment (ronniel1) [reply tools/generate_events.py:143]: You are right. 
Didn;t think of it.
I think it is better to have the  full ID in the events.yaml and just verify that there are no duplicates, in the generation tool.
What do you think?

Comment (ronniel1) [reply tools/generate_events.py]: fixed

Comment (avishayt) [reply tools/generate_events.py]: What is the ID for?  It still seems hard to maintain.  Can we just use name for ID?

**Comment by ybettan** [comment]
Can you please add the explanation you sent in slack in the commit message?

File: tools/generate_events.py (hunk 7 above)
Comment (ronniel1) [reply tools/generate_events.py]: ID can be whatever we decide.
The code generation script verifies that the ID is unique
I just started some (dum) convention for the ID names, but we can decide on something that makes sense. Only requirement is that it is unique

File: events.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    message: "Failed to cancel installation: error starting DB transaction"
+    event_type: cluster
+    severity: "error"
+    properties:
+        cluster_id:  UUID
+
+  - name: cancel_install_failed_commit
+    id: "CIE00002"
+    message: "Failed to cancel installation: error committing DB transaction"
+    event_type: cluster
```
Comment (gamli75) [inline events.yaml:13]: what are the options for the event_type? is it an enum?

**Review by gamli75 — COMMENTED** [review]
Do we want a different event for success/ failure? I mean 2 events: installation finished successfully & installation failed or one event with a different result
@avishayt

File: events.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
+x-events:
+  - name: cancel_install_failed_start
+    id: "CIE00001"
```
Comment (omertuc) [inline events.yaml:3]: What's the distinction between `CIE*` and `GIF*`?

File: events.yaml (hunk 8 above)
Comment (ronniel1) [reply events.yaml:13]: event type is used for validation.
And events code is generated slightly different

File: events.yaml (hunk 9 above)
Comment (ronniel1) [reply events.yaml:3]: thes are just some dumb convention I used. CIE - Cluster Install Error GIF - Generate Image Failure
Of course these are just examples we can use whatever convention we want for the IDs

**Comment by ronniel1** [comment]
> Can you please add the explanation you sent in slack in the commit message?

Good idea. Done

File: events.yaml (hunk 9 above)
Comment (tsorya) [reply events.yaml:3]: why we need id? sounds like name is enough no?

**Comment by ronniel1** [comment]
Closing this, as this will be handled by https://github.com/openshift/assisted-service/pull/1854

**Pull request closed without merging**